### DISCUSSION TITLE
Interaction with an unrelated library causes unexpected behavior of `tms.tiles(...)`

### DISCUSSION THREAD

**Discussion by bmcandr** [developmentseed/morecantile#138] [General]
I'll preface my comment by saying that the behavior I describe below does not occur in v5+ so it doesn't seem to be a critical issue. I'm just reporting something odd that I don't fully understand.

----

I've stumbled upon unusual behavior that appears to result from the interaction between `morecantile` and the [datadog-api-client library](https://github.com/DataDog/datadog-api-client-python). TLDR; `tms.tiles(...)` returns tiles outside the supplied bounds when modules from `datadog-api-client` are imported before `morecantile`.

We're running the following versions:

```
morecantile==4.3.0
datadog-api-client==2.17.0
```

We're still running 4.3.0 because we have not migrated our project to Pydantic v2 yet.

Example code:

```python
import json
import shapely.geometry
from datadog_api_client.v2 import ApiClient, ApiException, Configuration
import morecantile

feat = {
            "type": "Feature",
            "geometry": {
                "coordinates": [
                    [
                        [-98.547359, 29.522354],
                        [-98.547359, 29.335023],
                        [-98.27275, 29.335023],
                        [-98.27275, 29.522354],
                        [-98.547359, 29.522354],
                    ]
                ],
                "type": "Polygon",
            },
            "properties": {},
        }

tms = morecantile.tms.get("WebMercatorQuad")

g1 = shapely.geometry.shape(feat["geometry"])

west, south, east, north = g1.bounds

features = [tms.feature(tile) for tile in tms.tiles(west, south, east, north, zooms=[10], truncate=False)]

with open("tiles.geojson", "w") as f:
    json.dump(dict(type="FeatureCollection", features=features), f)
```

The above code produces a FeatureCollection containing tiles beginning at `x=0` rather than only the tiles that intersect `feat` (shown in red near San Antonio):
![Screenshot 2023-12-07 at 10 25 23](https://github.com/developmentseed/morecantile/assets/19274445/bd148bcf-0cb7-449a-99af-e22d51be4a87)

If I move the `morecantile` import before the `datadog-api-client` import or comment out the line importing modules from `datadog_api_client.v2`, only the tiles that intersect `feat` are returned as expected:
![Screenshot 2023-12-07 at 10 26 50](https://github.com/developmentseed/morecantile/assets/19274445/f673f436-ba9e-4460-98fc-0c345c68ab14)

Any idea why this happens? I first encountered this in our project repo and verified with a fresh venv with only the libraries required to run the snippet above.

**Comment by vincentsarago** [comment]
This is SOOOO interesting 

For some reason the pydantic model is altered by the import 

<img width="731" alt="Screenshot 2023-12-07 at 7 18 10 PM" src="https://github.com/developmentseed/morecantile/assets/10407788/9491e158-3c39-482a-be57-152727884af6">

See how the `union` order is different, in the first (good) case it's `number` then `integer` and in the second case (bad) it's `integer` then `number` 

The type is defined https://github.com/developmentseed/morecantile/blob/4.3.0/morecantile/models.py#L49

**Reply by bmcandr** [reply]
Weird! I confirm that there's something going on between Pydantic and Datadog that was revealed through our use of morecantile.

```python
# from datadog_api_client.v2 import ApiClient, ApiException, Configuration
from typing import Union, Tuple
from pydantic import BaseModel

NumType = Union[float, int]
BoundsType = Tuple[NumType, NumType]

class Example(BaseModel):
    b: BoundsType
    
e = Example(b=(-1.0, 1))

print(e.schema()["properties"]["b"]["items"])
```

Without datadog: `[{'anyOf': [{'type': 'number'}, {'type': 'integer'}]}, {'anyOf': [{'type': 'number'}, {'type': 'integer'}]}]`

With datadog: `[{'anyOf': [{'type': 'integer'}, {'type': 'number'}]}, {'anyOf': [{'type': 'integer'}, {'type': 'number'}]}]`

I'll report it to the Pydantic devs to see if they have any ideas. Thanks for taking a look!

**Reply by bmcandr** [reply]
Looks like it may actually be related to [this discussion on Python's runtime type caching](https://github.com/python/cpython/pull/103541#issuecomment-1517839464). My example can be further reduced to demonstrate the effect with just type aliases. Also, it only seems to arise with a `Tuple` of `Union`s? :shrug:

```python
from typing import List, Tuple, Union
from datadog_api_client.v2 import ApiClient, ApiException, Configuration

NumType = Union[float, int]
BoundsType = Tuple[NumType, NumType]
BoundsListType = List[NumType]

print(NumType)
print(BoundsType)
print(BoundsListType)

>>> typing.Union[float, int] # right
>>> typing.Tuple[typing.Union[int, float], typing.Union[int, float]]  # wrong
>>> typing.List[typing.Union[float, int]]  # right
```
